runtime/compiler/env/PersistentAllocator.cpp

```diff
@@ -68,78 +65,133 @@ PersistentAllocator::allocate(size_t size, void *hint)
+   {
+   Block * block = NULL;
+   Block * prev = NULL;
+   for (block = _freeBlocks[0]; block && block->_size < allocSize; prev = block, block = prev->next())
+      {}
+
+   if (block)
+      {
+      if (prev)
+         prev->_next = block->next();
+      else
+         _freeBlocks[0] = block->next();
+
+      block->_next = NULL;
+      }
+   return block;
+   }
+
 void *
-PersistentAllocator::allocateLocked(size_t requestedSize)
+PersistentAllocator::allocateInternal(size_t requestedSize)
    {
    TR_ASSERT( sizeof(Block) == mem_round( sizeof(Block) ),"Persistent block size will prevent us from properly aligning allocations.");
    size_t const dataSize = mem_round(requestedSize);
    size_t const allocSize = sizeof(Block) + dataSize;
 
-   TR::AllocatedMemoryMeter::update_allocated(allocSize, persistentAlloc);
+   if (TR::AllocatedMemoryMeter::_enabled & persistentAlloc)
+      {
+      // Use _smallBlockListsMonitor to protect TR::AllocatedMemoryMeter::update_allocated
+      // because accessing the variable-size-block list protected by ::memoryAllocMonitor
+      // takes longer and we may be penalizing access to fixed size block which should be very fast
+      j9thread_monitor_enter(_smallBlockListsMonitor);
+      TR::AllocatedMemoryMeter::update_allocated(allocSize, persistentAlloc);
+      j9thread_monitor_exit(_smallBlockListsMonitor);
+      }
 
    // If this is a small block try to allocate it from the appropriate
    // fixed-size-block chain.
    //
    size_t const index = freeBlocksIndex(allocSize);
-   Block * block = 0;
-   Block * prev = 0;
-   for (
-      block = _freeBlocks[index];
-      block && block->_size < allocSize;
-      prev = block, block = prev->next()
-      )
-      { TR_ASSERT(index == 0, "Iterating through a fixed-size block bin."); }
-
-   if (block)
+   if (index != 0) // fixed-size-block chain
       {
-
-      TR_ASSERT(
-         ( index == 0 ) || ( block->_size == allocSize ),
-         "block %p in chain for index %d has size %d (not %d)\n",
-         block,
-         index,
-         block->_size,
-         (index * sizeof(void *)) + sizeof(Block)
-         );
-
-      if (prev)
-         prev->_next = block->next();
-      else
+      j9thread_monitor_enter(_smallBlockListsMonitor);
+      Block * block = _freeBlocks[index];
+      if (block)
+         {
          _freeBlocks[index] = block->next();
-
-      block->_next = NULL;
-
-      size_t const excess = block->_size - allocSize;
-
-      if (excess > sizeof(Block))
+         block->_next = NULL;
+         j9thread_monitor_exit(_smallBlockListsMonitor);
+         return block + 1; // Return pointer after the header
+         }
+      j9thread_monitor_exit(_smallBlockListsMonitor);
+      // Need to allocate from segment
+      // Preemtively acquire the mutex that protects the segment allocation
+      if (::memoryAllocMonitor)
+         ::memoryAllocMonitor->enter();
```

Comment:
As explained above `::CriticalSection(::memoryAllocMonitor)` will not work because `::memoryAllocMonitor` can be NULL.
Another goal of this implementation was to reduce the number of monitor->enter()/monitor->exit() as much as possible. Otherwise I wouldn't rush into acquiring the monitor at this point

Reply:
Given that this is at the end of the `if` block, why does the monitor need to be acquired here? Wouldn't it be better to just move it right before the call to `allocateFromSegmentLocked` below? The `else` block looks self contained, so the only way to end up in the call to `allocateFromSegmentLocked` is if the code enter the `if` block here.

Reply:
When we try to allocate from the variable-size-block list and fail we have the corresponding lock in hand. The next action is to allocate from segments, operation which is protected by the same lock. Instead of releasing the lock just to acquire it immediately when segment allocation needs to happen, I decided to keep it on.  So we enter the section of code that allocates from segments with the monitor already held. This means that if we fail to allocate from the small-block list (the code above) we also need to lock the variable-size-block mutex in preparation of allocating from a segment.
The pseudo code looks like this:
```
If (need to allocate a small block)
   {
    try to allocate small block
    if (success) return
    else acquire LargeBlockMutex in prepartion for segment allocation
    }
else // Large block allocation
   {
    acquire LargeBlockMutex
    try to allocate large block
    if (success)   release mutex and return
     else
         {
        // LargeBlockMutex is still held
        // Do not release it just to acquire it again immediately
         }
     }
  AllocateFromSegment()
  release LargeBlockMutex
```

Reply:
If it feels cleaner I could repeat the code that does AllocateFromSegment into the two cases above (one for small blocks and one for large block)

Reply:
Ah I see your point. Yeah I think this is one of those few places where code duplication is a good idea. It would only be a duplication of
```
   allocation = allocateFromSegmentLocked(allocSize);
   if (::memoryAllocMonitor)
      ::memoryAllocMonitor->exit();
```
in both the `if` and `else` blocks, so I think the duplication is well worth the increased clarity it would provide.